tests/console/java.pm

```diff
@@ -21,56 +21,44 @@ use base "consoletest";
 use strict;
 use warnings;
 use testapi;
-use utils;
+use utils qw(pkcon_quit zypper_call);
+use version_utils 'is_sle';
 
 sub run {
-    select_console 'root-console';
-
+    my ($self) = @_;
+    $self->select_serial_terminal;
     # Make sure that PackageKit is not running
     pkcon_quit;
 
-    if (check_var("DISTRI", "sle")) {
-        if (zypper_call('in --auto-agree-with-licenses java-*', timeout => 500, exitcode => [0, 4])) {
-            # install only java-11-openjdk* & java-*-ibm*
-            zypper_call('in --auto-agree-with-licenses java-11-openjdk* java-*-ibm*');
-        }
-    }
-
-    if (check_var("DISTRI", "opensuse")) {
-        # Capture the return code value of the following scenarios
-        my $bootstrap_pkg_rt       = zypper_call("se java-*bootstrap",                             exitcode => [0, 104]);
-        my $bootstrap_conflicts_rt = zypper_call("in --auto-agree-with-licenses --dry-run java-*", exitcode => [0, 4]);
-
-        # logs / debugging purposes
-        diag "checking variable: bootstrap_pkg_rt = $bootstrap_pkg_rt";
-        diag "checking variable: bootstrap_conflicts_rt = $bootstrap_conflicts_rt";
-
-        if ($bootstrap_pkg_rt == 0) {
-            diag "There are java bootstrap packages available to be installed";
-            print "There are java bootstrap packages available to be installed\n";
-            if ($bootstrap_conflicts_rt == 0) {
-                diag "There is no conflict installing the java bootstrap packages";
-                print "There is no conflict installing the java bootstrap packages\n";
-                zypper_call("in java-*");
+    if (my $rc = zypper_call('in --auto-agree-with-licenses java-*', exitcode => [0, 4, 107])) {
+        # Workaround bsc#1115375, allow script failures
+        if ($rc == 107) {
+            unless (script_run('grep "java-9-openjdk-javadoc.*scriptlet failed" /var/log/zypp/history')) {
```

Comment:
This would also break if some other scriptlet fails in addition. I'd just keep the fail and not try to detect the softfail for this..

Reply:
the `else` belongs to this `unless`. So we should be covered, mainly the whole handling of 107 is part of this softfail code. When everything works fine, we should handle only the error code 4 which represents softfail [https://bugzilla.suse.com/show_bug.cgi?id=1137466](Bug 1137466 - [Build 20190606-1] sle15sp{0,1} pull obsoleted java-10 packages).

Reply:
> the else belongs to this unless.

Yes, but `grep` succeeds if the scriptlet failed and doesn't look at other failures.

> When everything works fine, we should handle only the error code 4

`4` means `ZYPPER_EXIT_ERR_ZYPP` and must not be taken as meaning success in any case.

The proper way for testing this is to explicitly exclude the obsolete JRE 10.

Reply:
soft fail only in case of 4 or 107 returned from zypper and each of them has to match specific recorded failure

Reply:
The match for `107` isn't specific enough though. If the log contains a mention of the grepped line anywhere, it accepts it.

IMO a softfail for this is just not worth it.

For the `4` case it's ok, because there's an explicit retry.

Reply:
> The match for `107` isn't specific enough though. If the log contains a mention of the grepped line anywhere, it accepts it.
> 
> IMO a softfail for this is just not worth it.
> 
I assume ( I have grepped tests and lib folders in os-autoinst-distri-opensuse )  that `java.pm` is the only intended java greedy installer. Otherwise `zypper in java` should download java-11. Maybe I am too much optimistic, but it shouldn't be pulled anywhere else. Maybe the package should be even deprecated since I haven't noticed it in sle12 during VRs.

Reply:
and with respect to https://bugzilla.suse.com/show_bug.cgi?id=1137466 claims that java-10 is obsoleted.

Reply:
Yes, but my point is that if later other java scriptlets also start to fail in addition to `java-9-openjdk-javadoc`, that would go entirely unnoticed. This scenario isn't too unlikely, as fixes get lost and regressions appear quite often.

Reply:
Yup and then it should be the default case unless the line was written in the log before `java.pm` execution as we discussed.

https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/b00b1c936621ffad5e75b0dfab0d8110004400cd/tests/console/java.pm#L49

or let's change grep to be more java greedy, 
`grep "java.*scriptlet failed"` 

It should not last there much. I hope that it will get fixed earlier if the problem will be highlighted.

Reply:
I think we're misunderstanding each other...

I'll try to explain my concern in a different way. Currently this does something like:
`zypper_call.retcode == 107 && zypper_call.failures.contains(java-9-javadoc scriptlet failed)`
while it should be
`zypper_call.retcode == 107 && zypper_call.failures == [java-9-javadoc scriptlet failed]`

IMO this case is too rare and the softfail was not necessary until now either, so I'd just drop it.

Reply:
> I think we're misunderstanding each other...
> 
> I'll try to explain my concern in a different way. Currently this does something like:
> `zypper_call.retcode == 107 && zypper_call.failures.contains(java-9-javadoc scriptlet failed)`
> while it should be
> `zypper_call.retcode == 107 && zypper_call.failures == [java-9-javadoc scriptlet failed]`
> 
> IMO this case is too rare and the softfail was not necessary until now either, so I'd just drop it.

But isn't the problem that the image pulls java-9 in fact? Because java-9 is not pulled by sle12sp{4,5} even. 
Checking scriptlet failures is not the proper detection, I admit it.

Reply:
The image does not pull it, the openQA test does.